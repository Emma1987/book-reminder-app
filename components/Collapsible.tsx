import { PropsWithChildren, useEffect, useRef, useState } from 'react';
import { Animated, Easing, StyleSheet, TouchableOpacity, View, useAnimatedValue } from 'react-native';
import Ionicons from '@expo/vector-icons/Ionicons';

import { ThemedText } from '@/components/ThemedText';
import { Colors } from '@/constants/Colors';

type CollapsibleProps = PropsWithChildren<{
    title: string;
    isCollapsibleOpen: boolean;
}>;

export function Collapsible({ children, title, isCollapsibleOpen = false }: CollapsibleProps) {
    const [isOpen, setIsOpen] = useState(isCollapsibleOpen);
    const animatedHeight = useAnimatedValue(isCollapsibleOpen ? 1 : 0);
    const [contentHeight, setContentHeight] = useState(0);

    useEffect(() => {
        Animated.timing(animatedHeight, {
            toValue: isOpen ? 1 : 0,
            duration: 300,
            easing: Easing.linear,
            useNativeDriver: false,
        }).start();
    }, [isOpen, contentHeight]);

    useEffect(() => {
        setIsOpen(isCollapsibleOpen);
    }, [isCollapsibleOpen]);

    const toggle = () => setIsOpen(!isOpen);

    return (
        <View>
            <TouchableOpacity
                style={styles.heading}
                onPress={toggle}
                activeOpacity={0.8}
                accessibilityRole="button"
                accessibilityState={{ expanded: isOpen }}
                accessibilityLabel={`${title} section`}
            >
                <Ionicons
                    name={isOpen ? 'chevron-down' : 'chevron-forward-outline'}
                    size={18}
                    color={Colors.icon}
                />
                <ThemedText type="defaultSemiBold">{title}</ThemedText>
            </TouchableOpacity>
            <Animated.View
                style={[
                    styles.content,
                    {
                        height: animatedHeight.interpolate({
                            inputRange: [0, 1],
                            outputRange: [0, contentHeight || 0],
                        }),
                    },
                ]}
            >
                <View
                    style={styles.animatedView}
                    onLayout={(event) => {
                        const measuredHeight = event.nativeEvent.layout.height;
                        setContentHeight(measuredHeight);
                    }}
                >
                    {children}
                </View>
            </Animated.View>
        </View>
    );
}

const styles = StyleSheet.create({
    heading: {
        flexDirection: 'row',
        alignItems: 'center',
        gap: 6,
    },
    content: {
        marginTop: 6,
        overflow: 'hidden',
    },
    animatedView: {
        minHeight: 1
    },
});
