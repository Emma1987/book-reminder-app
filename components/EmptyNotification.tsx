import React from 'react';
import { View, Image, StyleSheet } from 'react-native';

import { ThemedText } from '@/components/ThemedText';

export function EmptyNotification() {
    return (
        <View style={styles.container}>
            <Image
                source={require('@/assets/images/no-notification.png')}
                style={styles.image}
                resizeMode="contain"
                accessibilityRole="image"
                accessibilityLabel="No notification"
            />
            <ThemedText type="default" textAlign="center">
                No notification yet!
            </ThemedText>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        marginVertical: 50,
        alignItems: 'center',
    },
    image: {
        height: 200,
    },
});
