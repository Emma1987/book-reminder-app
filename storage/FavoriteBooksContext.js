import React, { createContext, useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const FavoriteBooksContext = createContext();

export const FavoriteBooksProvider = ({ children }) => {
    const [favorites, setFavorites] = useState([]);

    useEffect(() => {
        loadFavorites();
    }, []);

    const loadFavorites = async () => {
        const storedFavorites = await AsyncStorage.getItem('favorites');
        if (storedFavorites) setFavorites(JSON.parse(storedFavorites));
    };

    const addFavorite = async (book) => {
        const updatedFavorites = [...favorites, book];
        setFavorites(updatedFavorites);
        await AsyncStorage.setItem('favorites', JSON.stringify(updatedFavorites));
    };

    const removeFavorite = async (bookId) => {
        const updatedFavorites = favorites.filter((book) => book.id !== bookId);
        setFavorites(updatedFavorites);
        await AsyncStorage.setItem('favorites', JSON.stringify(updatedFavorites));
    };

    const isFavorite = (bookId) => favorites.some((book) => book.id === bookId);

    return (
        <FavoriteBooksContext.Provider value={{ favorites, addFavorite, removeFavorite, isFavorite }}>
            {children}
        </FavoriteBooksContext.Provider>
    );
};
