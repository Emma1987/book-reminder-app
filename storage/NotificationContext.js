import React, { createContext, setTimeout, useEffect, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as Notifications from 'expo-notifications';

export const NotificationContext = createContext();

export const NotificationProvider = ({ children }) => {
    const [unreadNotifications, setUnreadNotifications] = useState(0);
    const [notifications, setNotifications] = useState([]);

    const sortAndSetNotifications = (notifications) => {
        const sortedNotifications = notifications.sort((a, b) => new Date(b.scheduledTime) - new Date(a.scheduledTime));
        setNotifications(sortedNotifications);
    };

    const countUnreadNotifications = (notifications) => {
        const now = new Date();
        return notifications.filter((notification) => {
            return new Date(notification.scheduledTime) <= now && !notification.readAt;
        }).length;
    };

    useEffect(() => {
        loadNotifications();
    }, []);

    useEffect(() => {
        const unreadCount = countUnreadNotifications(notifications);
        setUnreadNotifications(unreadCount);
    }, [notifications]);

    useEffect(() => {
        const subscription = Notifications.addNotificationReceivedListener(async (notification) => {
            setTimeout(async () => {
                await loadNotifications();
            }, 5000);
        });

        return () => subscription.remove();
    }, []);

    const loadNotifications = async () => {
        try {
            const storedNotifications = await AsyncStorage.getItem('notifications');

            if (storedNotifications) {
                const parsedNotifications = JSON.parse(storedNotifications);
                sortAndSetNotifications([...parsedNotifications]);
            }
        } catch (error) {
            console.error('Error loading notifications:', error);
        }
    };

    const addNotification = async (notification) => {
        const updatedNotifications = [...notifications, notification];
        sortAndSetNotifications(updatedNotifications);
        await AsyncStorage.setItem('notifications', JSON.stringify(updatedNotifications));
    };

    const updateNotification = async (notificationId, updatedFields) => {
        const updatedNotifications = notifications.map((notification) =>
            notification.id === notificationId ? { ...notification, ...updatedFields } : notification,
        );

        sortAndSetNotifications(updatedNotifications);
        await AsyncStorage.setItem('notifications', JSON.stringify(updatedNotifications));
    };

    const removeNotification = async (notificationId) => {
        const updatedNotifications = notifications.filter((notification) => notification.id !== notificationId);
        sortAndSetNotifications(updatedNotifications);
        await AsyncStorage.setItem('notifications', JSON.stringify(updatedNotifications));
    };

    const getNotificationByBookId = (bookId) => notifications.find((notification) => notification.bookId === bookId);

    return (
        <NotificationContext.Provider
            value={{
                notifications,
                unreadNotifications,
                loadNotifications,
                addNotification,
                updateNotification,
                removeNotification,
                getNotificationByBookId,
            }}
        >
            {children}
        </NotificationContext.Provider>
    );
};
