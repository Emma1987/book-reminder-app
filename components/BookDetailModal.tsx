import React, { useMemo, useContext, useRef } from 'react';
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import BottomSheet, { BottomSheetScrollView } from '@gorhom/bottom-sheet';
import { Ionicons } from '@expo/vector-icons';

import { Separator } from '@/components/Separator';
import { ThemedText } from '@/components/ThemedText';
import BookType from '@/components/types';
import { Colors } from '@/constants/Colors';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';

export function BookDetailModal({ visible, onClose, book }) {
    const { addFavorite, removeFavorite, isFavorite } = useContext(FavoriteBooksContext);
    const bottomSheetRef = useRef<BottomSheet>(null);
    const snapPoints = useMemo(() => ['50%', '90%'], []);

    const closeModal = () => {
        onClose();
        bottomSheetRef.current.close();
    }

    const toggleFavorite = () => {
        if (isFavorite(book.id)) {
            removeFavorite(book.id);
        } else {
            addFavorite(book);
        }
    };

    return (
        <BottomSheet
            ref={bottomSheetRef}
            index={visible ? 0 : -1}
            snapPoints={snapPoints}
            onClose={closeModal}
            enablePanDownToClose={true}
            enableDynamicSizing={false}
        >
            <BottomSheetScrollView style={styles.bottomSheetView}>
                {/* Book Cover */}
                {book.coverImage &&
                    <Image source={{ uri: book.coverImage }} style={styles.bookCover} />
                }

                {/* Title, author and publication date */}
                <ThemedText type="subtitle-center">{book.title}</ThemedText>
                <ThemedText type="defaultSemiBold-center" style={styles.mb10}>Author: {book.authors?.join(', ')}</ThemedText>
                <ThemedText type="info-center" style={styles.mb10}>Publication date: {book.publicationDateStr}</ThemedText>

                {/* Language */}
                {book.language && 
                    <ThemedText type="info-center" style={styles.mb10}>Language: {book.language.toUpperCase()}</ThemedText>
                }

                {/* ISBN */}
                {book.isbn10 && <ThemedText type="info-center">ISBN10: {book.isbn10}</ThemedText>}
                {book.isbn13 && <ThemedText type="info-center">ISBN13: {book.isbn13}</ThemedText>}

                {/* Description */}
                {book.description &&
                    <View>
                        <Separator />
                        <Text style={styles.description}>{book.description}</Text>
                    </View>
                }

                {/* Add to favs Button */}
                {isFavorite(book.id) && 
                    <TouchableOpacity style={[styles.button, styles.removeFavButton]} onPress={toggleFavorite}>
                        <Text style={styles.buttonText}>Remove from favorites</Text>
                    </TouchableOpacity>
                }
                {!isFavorite(book.id) && 
                    <TouchableOpacity style={[styles.button, styles.addFavButton]} onPress={toggleFavorite}>
                        <Text style={styles.buttonText}>Add to favorites</Text>
                    </TouchableOpacity>
                }
            </BottomSheetScrollView>
        </BottomSheet>
    );
}

const styles = StyleSheet.create({
    bottomSheetView: {
        padding: 20,
    },
    bookCover: {
        width: 100,
        height: 150,
        alignSelf: 'center',
        marginBottom: 16,
    },
    textCenter: {
        textAlign: 'center'
    },
    mb10: {
        marginBottom: 10,
    },
    description: {
        fontSize: 14,
        textAlign: 'justify',
        marginBottom: 50,
    },
    button: {
        paddingVertical: 10,
        paddingHorizontal: 20,
        marginBottom: 50,
        borderRadius: 5,
        width: 150,
        alignSelf: 'center',
    },
    buttonText: {
        color: Colors.white,
        fontWeight: 'bold',
        textAlign: 'center',
    },
    addFavButton: {
        backgroundColor: Colors.red,
    },
    removeFavButton: {
        backgroundColor: Colors.gray,
    },
});
