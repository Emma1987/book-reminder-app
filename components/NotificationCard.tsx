import React, { useContext, useEffect, useState } from 'react';
import { Image, StyleSheet, Text, View } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';

import { ThemedText } from '@/components/ThemedText';
import { NotificationType } from '@/components/types';
import { Colors } from '@/constants/Colors';
import { timeAgo } from '@/helpers/DateHelper';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';

type NotificationCardProps = {
    notification: NotificationType;
};

export function NotificationCard({ notification }: NotificationCardProps) {
    const { getFavoriteById } = useContext(FavoriteBooksContext);

    const [timeAgoString, setTimeAgoString] = useState(timeAgo(notification.scheduledTime));

    const now = new Date();

    useEffect(() => {
        const interval = setInterval(() => {
            setTimeAgoString(timeAgo(notification.scheduledTime));
        }, 60000);

        return () => clearInterval(interval);
    }, [notification.scheduledTime]);

    const book = getFavoriteById(notification.bookId);
    if (new Date(notification.scheduledTime) > now || !book) {
        return;
    }

    return (
        <View style={styles.container}>
            <View>
                <Image
                    source={
                        book?.coverImage
                            ? { uri: book.coverImage }
                            : require('@/assets/images/book-cover-placeholder.jpg')
                    }
                    style={styles.bookCover}
                    accessibilityRole="image"
                    accessibilityLabel="Book cover"
                />
            </View>
            <View style={styles.notificationContent}>
                <View style={styles.title}>
                    <ThemedText type="defaultSemiBold">New release!</ThemedText>
                    <View style={styles.notificationDate}>
                        <ThemedText type="info">{timeAgoString}</ThemedText>
                        {!notification.readAt && (
                            <FontAwesome name="circle" size={14} color={Colors.red} style={styles.newNotification} />
                        )}
                    </View>
                </View>
                <Text>{notification.content}</Text>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        padding: 20,
        borderBottomWidth: 1,
        borderColor: Colors.white,
    },
    bookCover: {
        width: 40,
        height: 65,
        marginRight: 10,
    },
    notificationContent: {
        flex: 1,
    },
    title: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        paddingBottom: 5,
    },
    notificationDate: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    newNotification: {
        paddingLeft: 5,
    },
});
