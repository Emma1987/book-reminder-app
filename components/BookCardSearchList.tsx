import React, { useContext } from 'react';
import { Alert, Image, Keyboard, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

import { ThemedText } from '@/components/ThemedText';
import { BookType } from '@/components/types';
import { Colors } from '@/constants/Colors';
import { isPublished } from '@/helpers/BookHelper';
import { formatDateStr } from '@/helpers/DateHelper';
import { scheduleBookReleaseNotification } from '@/helpers/BookNotificationHelper';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';
import { NotificationContext } from '@/storage/NotificationContext';

type BookCardSearchListProps = {
    book: BookType;
    onSeeDetails: (book: BookType) => void;
};

export function BookCardSearchList({ book, onSeeDetails }: BookCardSearchListProps) {
    const { addFavorite, removeFavorite, isFavorite } = useContext(FavoriteBooksContext);
    const { addNotification } = useContext(NotificationContext);

    const getNotified = () => {
        const alertContent = `ðŸ“š The book "${book.title}" is set to be released on ${formatDateStr(book.releaseDateRaw, 'monthFirst')}.\n\nWould you like to receive a notification on the release date?`;

        Alert.alert('Stay Updated!', alertContent, [
            {
                text: 'No, thanks',
                style: 'cancel',
                onPress: () => console.log('Notification declined'),
            },
            {
                text: 'Yes, notify me!',
                style: 'default',
                onPress: async () => {
                    scheduleBookReleaseNotification(book, addNotification);
                },
            },
        ]);
    };

    const toggleFavorite = () => {
        Keyboard.dismiss();

        if (isFavorite(book.id)) {
            removeFavorite(book.id);
        } else {
            if (!isPublished(book) && book.releaseDateRaw !== null) {
                getNotified();
            }

            addFavorite(book);
        }
    };

    return (
        <View style={styles.container}>
            {/* Book Cover */}
            <Image
                source={
                    book.coverImage ? { uri: book.coverImage } : require('@/assets/images/book-cover-placeholder.jpg')
                }
                style={styles.bookCover}
                accessibilityRole="image"
                accessibilityLabel="Book cover"
            />

            {/* Book Description */}
            <View style={styles.bookDescription}>
                <ThemedText type="subtitle">{book.title}</ThemedText>
                <ThemedText type="defaultSemiBold">{book.authors?.[0] ?? ''}</ThemedText>
                <Text>Release date: {formatDateStr(book.releaseDateRaw)}</Text>
            </View>

            {/* Action Icons */}
            <View style={styles.actions}>
                <TouchableOpacity
                    onPress={toggleFavorite}
                    accessibilityLabel="Add to favorites"
                    accessibilityHint="Adds this book to your favorites list"
                >
                    {isFavorite(book.id) && <Ionicons name="heart" size={25} color={Colors.red} />}
                    {!isFavorite(book.id) && <Ionicons name="heart-outline" size={25} color={Colors.gray} />}
                </TouchableOpacity>
                <TouchableOpacity
                    onPress={() => onSeeDetails(book)}
                    accessibilityLabel="View book details"
                    accessibilityHint={`View more details about ${book.title}`}
                >
                    <Ionicons name="information-circle-outline" size={25} color={Colors.gray} />
                </TouchableOpacity>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        padding: 10,
        borderBottomWidth: 1,
        borderColor: Colors.white,
    },
    bookCover: {
        width: 60,
        height: 90,
    },
    bookDescription: {
        flex: 1,
        padding: 10,
    },
    actions: {
        width: 60,
        justifyContent: 'space-between',
        alignItems: 'center',
    },
});
