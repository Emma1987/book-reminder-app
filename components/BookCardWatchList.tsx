import React, { useContext } from 'react';
import { Alert, Image, StyleSheet, TouchableOpacity, View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

import { ThemedText } from '@/components/ThemedText';
import { BookType, NotificationType } from '@/components/types';
import { Colors } from '@/constants/Colors';
import { isPublished } from '@/helpers/BookHelper';
import { scheduleBookReleaseNotification } from '@/helpers/BookNotificationHelper';
import { formatDateStr } from '@/helpers/DateHelper';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';
import { NotificationContext } from '@/storage/NotificationContext';

type BookCardWatchListProps = {
    book: BookType;
};

export function BookCardWatchList({ book }: BookCardWatchListProps) {
    const { removeFavorite } = useContext(FavoriteBooksContext);
    const { addNotification, removeNotification, getNotificationByBookId } = useContext(NotificationContext);

    const notification: NotificationType | undefined = getNotificationByBookId(book.id);

    const getNotified = () => {
        const alertContent = `ðŸ“š The book "${book.title}" is set to be released on ${formatDateStr(book.releaseDateRaw, 'monthFirst')}.\n\nWould you like to receive a notification on the release date?`;

        Alert.alert('Stay Updated!', alertContent, [
            {
                text: 'No, thanks',
                style: 'cancel',
                onPress: () => console.log('Notification declined'),
            },
            {
                text: 'Yes, notify me!',
                style: 'default',
                onPress: async () => {
                    scheduleBookReleaseNotification(book, addNotification);
                },
            },
        ]);
    };

    const handleToggleNotification = () => {
        if (!notification) {
            getNotified();

            return;
        }

        removeNotification(notification.id);
    };

    const handleRemoveFavorite = () => {
        Alert.alert('Remove Favorite', `Are you sure you want to remove "${book.title}" from your favorites?`, [
            {
                text: 'Cancel',
                style: 'cancel',
                onPress: () => {},
            },
            {
                text: 'Remove',
                style: 'destructive',
                onPress: () => {
                    if (notification) {
                        removeNotification(notification.id);
                    }

                    removeFavorite(book.id);
                },
            },
        ]);
    };

    return (
        <View style={styles.container}>
            {/* Book Cover */}
            <Image
                source={
                    book.coverImage ? { uri: book.coverImage } : require('@/assets/images/book-cover-placeholder.jpg')
                }
                style={styles.bookCover}
                accessibilityRole="image"
                accessibilityLabel="Book cover"
            />

            {/* Book Description */}
            <View style={styles.bookDescription}>
                <ThemedText type="defaultSemiBold">{book.title}</ThemedText>
                <ThemedText type="default">{book.authors?.[0] ?? ''}</ThemedText>
                <ThemedText type="info">Release date: {formatDateStr(book.releaseDateRaw)}</ThemedText>
            </View>

            {/* Action Icons */}
            <View style={styles.actions}>
                {!isPublished(book) && (
                    <TouchableOpacity
                        onPress={handleToggleNotification}
                        accessibilityLabel="Notify me when the book is released"
                    >
                        <Ionicons
                            name={notification ? 'notifications' : 'notifications-off'}
                            size={20}
                            color={notification ? Colors.gold : Colors.gray}
                        />
                    </TouchableOpacity>
                )}
                <TouchableOpacity
                    onPress={handleRemoveFavorite}
                    accessibilityLabel="Remove this book from your favorite list"
                >
                    <Ionicons name="trash-outline" size={20} color={Colors.red} />
                </TouchableOpacity>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        padding: 10,
        backgroundColor: Colors.whitesmoke,
        borderRadius: 5,
        marginBottom: 5,
    },
    bookCover: {
        width: 40,
        height: 60,
    },
    bookDescription: {
        flex: 1,
        paddingLeft: 10,
    },
    actions: {
        width: 30,
        gap: 5,
        alignItems: 'center',
    },
});
