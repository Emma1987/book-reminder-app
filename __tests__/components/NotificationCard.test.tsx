import React from 'react';
import { render } from '@testing-library/react-native';
import { littleSecretsBook, littleSecretsBookNotification, veryAnticipatedBook, veryAnticipatedBookNotification } from '@/__tests__/__fixtures__/fixtures';
import { NotificationCard } from '@/components/NotificationCard';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';
import { NotificationContext } from '@/storage/NotificationContext';

jest.mock('@/helpers/DateHelper', () => ({
    timeAgo: jest.fn(() => '5 minutes ago'),
}));

describe('NotificationCard', () => {
    it('renders notification information correctly', () => {
        const { getByText } = render(
            <FavoriteBooksContext.Provider value={{getFavoriteById: () => littleSecretsBook}}>
                <NotificationContext.Provider value={{addNotification: jest.fn(), removeNotification: jest.fn(), getNotificationByBookId: jest.fn()}}>
                    <NotificationCard notification={littleSecretsBookNotification} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        expect(getByText('New release!')).toBeTruthy();
        expect(getByText('Little Secrets is released today!')).toBeTruthy();
        expect(getByText('5 minutes ago')).toBeTruthy();
        expect(getByText('circle')).toBeTruthy();
    });

    it('does not render if the book is not in favorites', () => {
        const { queryByText } = render(
            <FavoriteBooksContext.Provider value={{getFavoriteById: () => undefined}}>
                <NotificationContext.Provider value={{addNotification: jest.fn(), removeNotification: jest.fn(), getNotificationByBookId: jest.fn()}}>
                    <NotificationCard notification={littleSecretsBookNotification} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        expect(queryByText('New release!')).toBeNull();
    });

    it('returns null if notification is not scheduled to be sent yet', () => {
        const { queryByText } = render(
            <FavoriteBooksContext.Provider value={{getFavoriteById: () => veryAnticipatedBook}}>
                <NotificationContext.Provider value={{addNotification: jest.fn(), removeNotification: jest.fn(), getNotificationByBookId: jest.fn()}}>
                    <NotificationCard notification={veryAnticipatedBookNotification} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        expect(queryByText('New release!')).toBeNull();
    });

    it('does not show unread indicator if notification is read', () => {
        const readNotification = { ...littleSecretsBookNotification, readAt: new Date() };
        const { queryByText } = render(
            <FavoriteBooksContext.Provider value={{getFavoriteById: () => littleSecretsBook}}>
                <NotificationContext.Provider value={{addNotification: jest.fn(), removeNotification: jest.fn(), getNotificationByBookId: jest.fn()}}>
                    <NotificationCard notification={readNotification} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        expect(queryByText('circle')).toBeNull();
    });

    it('matches the snapshot', () => {
        const tree = render(
            <FavoriteBooksContext.Provider value={{getFavoriteById: () => littleSecretsBook}}>
                <NotificationContext.Provider value={{addNotification: jest.fn(), removeNotification: jest.fn(), getNotificationByBookId: jest.fn()}}>
                    <NotificationCard notification={littleSecretsBookNotification} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
