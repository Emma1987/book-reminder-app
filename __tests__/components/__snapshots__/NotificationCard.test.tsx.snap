// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationCard matches the snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderBottomWidth": 1,
      "borderColor": "white",
      "flexDirection": "row",
      "padding": 20,
    }
  }
>
  <View>
    <Image
      accessibilityLabel="Book cover"
      accessibilityRole="image"
      source={
        {
          "uri": "https://book.com/little-secrets-cover.jpg",
        }
      }
      style={
        {
          "height": 65,
          "marginRight": 10,
          "width": 40,
        }
      }
    />
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 5,
        }
      }
    >
      <Text
        style={
          [
            undefined,
            undefined,
            {
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        New release!
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              {
                "color": "gray",
                "fontSize": 14,
              },
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          5 minutes ago
        </Text>
        <Text>
          circle
        </Text>
      </View>
    </View>
    <Text>
      Little Secrets is released today!
    </Text>
  </View>
</View>
`;
